### PR TITLE
Feature/cursor improvments

### PR DESCRIPTION
**Pull request by lwasylow**
Few features improvements.
Detailed calculations are done on the subset of data after we establish which rows do not match using pk has.
Calculating include, exclude only once in early step so we dont repeat over and over, same for pk hash and value.

### CODE REVIEW

File: source/expectations/data_values/ut_compound_data_value.tpb [hunk 1]
```diff
@@ -229,63 +251,80 @@ create or replace type body ut_compound_data_value as
...
+    **/     
     execute immediate 'merge into ' || l_ut_owner || '.ut_compound_data_tmp tgt
                        using (
                               select '||l_ut_owner ||'.ut_compound_data_helper.get_hash(ucd.item_data.getclobval()) item_hash, 
-                                      pk_hash, ucd.item_no, ucd.data_id
+                                      ucd.item_no, ucd.data_id, ucd.item_data
                               from
                               (
-                              select '||l_column_filter||','||get_column_pk_hash(a_join_by_xpath)||', item_no, data_id
+                              select '||l_column_filter||', item_no, data_id
```
Comment (utplsql-travis) [inline source/expectations/data_values/ut_compound_data_value.tpb]: Code Smell Code Smell: An illegal character with code point 10 was found in this literal. ([plsql:LiteralsNonPrintableCharactersCheck](https://sonarcloud.io/organizations/utplsql/rules#rule_key=plsql%3ALiteralsNonPrintableCharactersCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=utPLSQL&pullRequest=760&issues=AWY_N4wenzfF1D4JsH2J&open=AWY_N4wenzfF1D4JsH2J)

[ISSUE_KEY]: AWY_N4wenzfF1D4JsH2J

File: source/expectations/data_values/ut_compound_data_value.tpb [hunk 2]
```diff
@@ -229,63 +251,80 @@ create or replace type body ut_compound_data_value as
...
+                        t.item_data,
+                        row_number() over (partition by t.item_hash order by 1) duplicate_no
+                        from  ' || l_ut_owner || '.ut_compound_data_tmp t
+                        where t.data_id = :self_guid
+                        ),
+                        expected as (
+                           select t.item_hash,
+                           t.item_data,
+                           row_number() over (partition by t.item_hash order by 1) duplicate_no
                            from  ' || l_ut_owner || '.ut_compound_data_tmp t
```
Comment (utplsql-travis) [inline source/expectations/data_values/ut_compound_data_value.tpb]: Code Smell Code Smell: An illegal character with code point 10 was found in this literal. ([plsql:LiteralsNonPrintableCharactersCheck](https://sonarcloud.io/organizations/utplsql/rules#rule_key=plsql%3ALiteralsNonPrintableCharactersCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=utPLSQL&pullRequest=760&issues=AWY_N4wenzfF1D4JsH2M&open=AWY_N4wenzfF1D4JsH2M)

[ISSUE_KEY]: AWY_N4wenzfF1D4JsH2M

File: source/expectations/data_values/ut_compound_data_value.tpb [hunk 3]
```diff
@@ -218,6 +218,28 @@ create or replace type body ut_compound_data_value as
       return l_column;
     end;
     
+    procedure produce_pk_hash(a_pk_hash_extr_string varchar2, a_join_by_xpath in varchar2, a_diff_id raw) is
+    begin
+      execute immediate 'merge into ' || l_ut_owner || '.ut_compound_data_diff_tmp tgt
```
Comment (utplsql-travis) [inline source/expectations/data_values/ut_compound_data_value.tpb]: Code Smell Code Smell: Add a WHEN OTHERS exception handler to ensure that all exceptions are trapped. ([plsql:ExecuteImmediateTrapExceptionsCheck](https://sonarcloud.io/organizations/utplsql/rules#rule_key=plsql%3AExecuteImmediateTrapExceptionsCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=utPLSQL&pullRequest=760&issues=AWY_N4wenzfF1D4JsH2F&open=AWY_N4wenzfF1D4JsH2F)

[ISSUE_KEY]: AWY_N4wenzfF1D4JsH2F

File: source/expectations/data_values/ut_compound_data_helper.pkb [hunk 4]
```diff
@@ -231,9 +228,9 @@ create or replace package body ut_compound_data_helper is
                   r.duplicate_no,
                   ucd.column_value.getclobval() col_val,
                   ucd.column_value.getRootElement() col_name,
-                  ut_compound_data_helper.get_pk_value(:join_xpath,r.item_data) pk_value
+                  r.pk_value
                   from 
-                   (select ]'||l_column_filter||q'[, ucd.item_no, ucd.item_hash, i.pk_hash, i.duplicate_no
+                   (select ]'||l_column_filter||q'[, ucd.item_no, ucd.item_hash, i.pk_hash, i.pk_value, i.duplicate_no
```
Comment (utplsql-travis) [inline source/expectations/data_values/ut_compound_data_helper.pkb]: Code Smell Code Smell: An illegal character with code point 10 was found in this literal. ([plsql:LiteralsNonPrintableCharactersCheck](https://sonarcloud.io/organizations/utplsql/rules#rule_key=plsql%3ALiteralsNonPrintableCharactersCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=utPLSQL&pullRequest=760&issues=AWY_N4w2nzfF1D4JsH2O&open=AWY_N4w2nzfF1D4JsH2O)

[ISSUE_KEY]: AWY_N4w2nzfF1D4JsH2O

File: source/expectations/data_values/ut_compound_data_value.tpb [hunk 5]
```diff
@@ -229,63 +251,80 @@ create or replace type body ut_compound_data_value as
...
+                       set tgt.item_hash = src.item_hash]'
+                       using a_exclude_xpath, a_include_xpath,
+                             self.data_id, l_other.data_id;
     
     /* Peform minus on two sets two get diffrences that will be used later on to print results */
-    execute immediate 'insert into ' || l_ut_owner || '.ut_compound_data_diff_tmp ( diff_id,item_hash,pk_hash,duplicate_no)
-                       with source_data as
-                       ( select t.data_id,t.item_hash,row_number() over (partition by t.pk_hash,t.item_hash,t.data_id order by 1,2) duplicate_no,
-                           pk_hash
+    execute immediate 'insert into ' || l_ut_owner || '.ut_compound_data_diff_tmp ( diff_id,item_hash,duplicate_no,item_data)
```
Comment (utplsql-travis) [inline source/expectations/data_values/ut_compound_data_value.tpb]: Code Smell Code Smell: An illegal character with code point 10 was found in this literal. ([plsql:LiteralsNonPrintableCharactersCheck](https://sonarcloud.io/organizations/utplsql/rules#rule_key=plsql%3ALiteralsNonPrintableCharactersCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=utPLSQL&pullRequest=760&issues=AWY_N4wenzfF1D4JsH2L&open=AWY_N4wenzfF1D4JsH2L)

[ISSUE_KEY]: AWY_N4wenzfF1D4JsH2L

File: source/expectations/data_values/ut_compound_data_value.tpb [hunk 6]
```diff
@@ -218,6 +218,28 @@ create or replace type body ut_compound_data_value as
       return l_column;
     end;
     
+    procedure produce_pk_hash(a_pk_hash_extr_string varchar2, a_join_by_xpath in varchar2, a_diff_id raw) is
+    begin
+      execute immediate 'merge into ' || l_ut_owner || '.ut_compound_data_diff_tmp tgt
+        using (
+          select '||a_pk_hash_extr_string ||', diff_id, item_hash, duplicate_no,
```
Comment (utplsql-travis) [inline source/expectations/data_values/ut_compound_data_value.tpb]: Code Smell Code Smell: An illegal character with code point 10 was found in this literal. ([plsql:LiteralsNonPrintableCharactersCheck](https://sonarcloud.io/organizations/utplsql/rules#rule_key=plsql%3ALiteralsNonPrintableCharactersCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=utPLSQL&pullRequest=760&issues=AWY_N4wenzfF1D4JsH2H&open=AWY_N4wenzfF1D4JsH2H)

[ISSUE_KEY]: AWY_N4wenzfF1D4JsH2H

File: source/expectations/data_values/ut_compound_data_value.tpb (hunk 5 above)
Comment (utplsql-travis) [inline source/expectations/data_values/ut_compound_data_value.tpb]: Code Smell Code Smell: Add a WHEN OTHERS exception handler to ensure that all exceptions are trapped. ([plsql:ExecuteImmediateTrapExceptionsCheck](https://sonarcloud.io/organizations/utplsql/rules#rule_key=plsql%3AExecuteImmediateTrapExceptionsCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=utPLSQL&pullRequest=760&issues=AWY_N4wenzfF1D4JsH2K&open=AWY_N4wenzfF1D4JsH2K)

[ISSUE_KEY]: AWY_N4wenzfF1D4JsH2K

File: source/expectations/data_values/ut_compound_data_helper.pkb [hunk 7]
```diff
@@ -211,9 +208,9 @@ create or replace package body ut_compound_data_helper is
                  r.duplicate_no,
                  ucd.column_value.getclobval() col_val,
                  ucd.column_value.getRootElement() col_name,
-                 ut_compound_data_helper.get_pk_value(:join_xpath,r.item_data) pk_value
+                 r.pk_value
                  from 
-                  (select ]'||l_column_filter||q'[, ucd.item_no, ucd.item_hash, i.pk_hash, i.duplicate_no
+                  (select ]'||l_column_filter||q'[, ucd.item_no, ucd.item_hash, i.pk_hash, i.pk_value, i.duplicate_no
```
Comment (utplsql-travis) [inline source/expectations/data_values/ut_compound_data_helper.pkb]: Code Smell Code Smell: An illegal character with code point 10 was found in this literal. ([plsql:LiteralsNonPrintableCharactersCheck](https://sonarcloud.io/organizations/utplsql/rules#rule_key=plsql%3ALiteralsNonPrintableCharactersCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=utPLSQL&pullRequest=760&issues=AWY_N4w2nzfF1D4JsH2N&open=AWY_N4w2nzfF1D4JsH2N)

[ISSUE_KEY]: AWY_N4w2nzfF1D4JsH2N

File: source/expectations/data_values/ut_compound_data_value.tpb (hunk 3 above)
Comment (utplsql-travis) [inline source/expectations/data_values/ut_compound_data_value.tpb]: Code Smell Code Smell: An illegal character with code point 10 was found in this literal. ([plsql:LiteralsNonPrintableCharactersCheck](https://sonarcloud.io/organizations/utplsql/rules#rule_key=plsql%3ALiteralsNonPrintableCharactersCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=utPLSQL&pullRequest=760&issues=AWY_N4wenzfF1D4JsH2G&open=AWY_N4wenzfF1D4JsH2G)

[ISSUE_KEY]: AWY_N4wenzfF1D4JsH2G

File: source/expectations/data_values/ut_compound_data_value.tpb [hunk 8]
```diff
@@ -218,6 +218,28 @@ create or replace type body ut_compound_data_value as
...
+          replace((extract(ucd.item_data,:join_by_xpath).getclobval()),chr(10)) pk_value
+          from ' || l_ut_owner || '.ut_compound_data_diff_tmp ucd
```
Comment (utplsql-travis) [inline source/expectations/data_values/ut_compound_data_value.tpb]: Code Smell Code Smell: An illegal character with code point 10 was found in this literal. ([plsql:LiteralsNonPrintableCharactersCheck](https://sonarcloud.io/organizations/utplsql/rules#rule_key=plsql%3ALiteralsNonPrintableCharactersCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=utPLSQL&pullRequest=760&issues=AWY_N4wenzfF1D4JsH2I&open=AWY_N4wenzfF1D4JsH2I)

[ISSUE_KEY]: AWY_N4wenzfF1D4JsH2I

File: source/expectations/data_values/ut_compound_data_value.tpb (hunk 2 above)
Comment (utplsql-travis) [inline source/expectations/data_values/ut_compound_data_value.tpb]: Code Smell Code Smell: An illegal character with code point 10 was found in this literal. ([plsql:LiteralsNonPrintableCharactersCheck](https://sonarcloud.io/organizations/utplsql/rules#rule_key=plsql%3ALiteralsNonPrintableCharactersCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=utPLSQL&pullRequest=760&issues=AWY_N4wenzfF1D4JsH2M&open=AWY_N4wenzfF1D4JsH2M)

[ISSUE_KEY]: AWY_N4wenzfF1D4JsH2M

File: source/expectations/data_values/ut_compound_data_value.tpb (hunk 5 above)
Comment (utplsql-travis) [inline source/expectations/data_values/ut_compound_data_value.tpb]: Code Smell Code Smell: Add a WHEN OTHERS exception handler to ensure that all exceptions are trapped. ([plsql:ExecuteImmediateTrapExceptionsCheck](https://sonarcloud.io/organizations/utplsql/rules#rule_key=plsql%3AExecuteImmediateTrapExceptionsCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=utPLSQL&pullRequest=760&issues=AWY_N4wenzfF1D4JsH2K&open=AWY_N4wenzfF1D4JsH2K)

[ISSUE_KEY]: AWY_N4wenzfF1D4JsH2K

File: source/expectations/data_values/ut_compound_data_value.tpb (hunk 6 above)
Comment (utplsql-travis) [inline source/expectations/data_values/ut_compound_data_value.tpb]: Code Smell Code Smell: An illegal character with code point 10 was found in this literal. ([plsql:LiteralsNonPrintableCharactersCheck](https://sonarcloud.io/organizations/utplsql/rules#rule_key=plsql%3ALiteralsNonPrintableCharactersCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=utPLSQL&pullRequest=760&issues=AWY_N4wenzfF1D4JsH2H&open=AWY_N4wenzfF1D4JsH2H)

[ISSUE_KEY]: AWY_N4wenzfF1D4JsH2H

File: source/expectations/data_values/ut_compound_data_value.tpb (hunk 3 above)
Comment (utplsql-travis) [inline source/expectations/data_values/ut_compound_data_value.tpb]: Code Smell Code Smell: Add a WHEN OTHERS exception handler to ensure that all exceptions are trapped. ([plsql:ExecuteImmediateTrapExceptionsCheck](https://sonarcloud.io/organizations/utplsql/rules#rule_key=plsql%3AExecuteImmediateTrapExceptionsCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=utPLSQL&pullRequest=760&issues=AWY_N4wenzfF1D4JsH2F&open=AWY_N4wenzfF1D4JsH2F)

[ISSUE_KEY]: AWY_N4wenzfF1D4JsH2F

File: source/expectations/data_values/ut_compound_data_helper.pkb (hunk 7 above)
Comment (utplsql-travis) [inline source/expectations/data_values/ut_compound_data_helper.pkb]: Code Smell Code Smell: An illegal character with code point 10 was found in this literal. ([plsql:LiteralsNonPrintableCharactersCheck](https://sonarcloud.io/organizations/utplsql/rules#rule_key=plsql%3ALiteralsNonPrintableCharactersCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=utPLSQL&pullRequest=760&issues=AWY_N4w2nzfF1D4JsH2N&open=AWY_N4w2nzfF1D4JsH2N)

[ISSUE_KEY]: AWY_N4w2nzfF1D4JsH2N

File: source/expectations/data_values/ut_compound_data_value.tpb (hunk 5 above)
Comment (utplsql-travis) [inline source/expectations/data_values/ut_compound_data_value.tpb]: Code Smell Code Smell: An illegal character with code point 10 was found in this literal. ([plsql:LiteralsNonPrintableCharactersCheck](https://sonarcloud.io/organizations/utplsql/rules#rule_key=plsql%3ALiteralsNonPrintableCharactersCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=utPLSQL&pullRequest=760&issues=AWY_N4wenzfF1D4JsH2L&open=AWY_N4wenzfF1D4JsH2L)

[ISSUE_KEY]: AWY_N4wenzfF1D4JsH2L

File: source/expectations/data_values/ut_compound_data_value.tpb (hunk 8 above)
Comment (utplsql-travis) [inline source/expectations/data_values/ut_compound_data_value.tpb]: Code Smell Code Smell: An illegal character with code point 10 was found in this literal. ([plsql:LiteralsNonPrintableCharactersCheck](https://sonarcloud.io/organizations/utplsql/rules#rule_key=plsql%3ALiteralsNonPrintableCharactersCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=utPLSQL&pullRequest=760&issues=AWY_N4wenzfF1D4JsH2I&open=AWY_N4wenzfF1D4JsH2I)

[ISSUE_KEY]: AWY_N4wenzfF1D4JsH2I

File: source/expectations/data_values/ut_compound_data_value.tpb (hunk 3 above)
Comment (utplsql-travis) [inline source/expectations/data_values/ut_compound_data_value.tpb]: Code Smell Code Smell: An illegal character with code point 10 was found in this literal. ([plsql:LiteralsNonPrintableCharactersCheck](https://sonarcloud.io/organizations/utplsql/rules#rule_key=plsql%3ALiteralsNonPrintableCharactersCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=utPLSQL&pullRequest=760&issues=AWY_N4wenzfF1D4JsH2G&open=AWY_N4wenzfF1D4JsH2G)

[ISSUE_KEY]: AWY_N4wenzfF1D4JsH2G

File: source/expectations/data_values/ut_compound_data_value.tpb (hunk 1 above)
Comment (utplsql-travis) [inline source/expectations/data_values/ut_compound_data_value.tpb]: Code Smell Code Smell: An illegal character with code point 10 was found in this literal. ([plsql:LiteralsNonPrintableCharactersCheck](https://sonarcloud.io/organizations/utplsql/rules#rule_key=plsql%3ALiteralsNonPrintableCharactersCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=utPLSQL&pullRequest=760&issues=AWY_N4wenzfF1D4JsH2J&open=AWY_N4wenzfF1D4JsH2J)

[ISSUE_KEY]: AWY_N4wenzfF1D4JsH2J

File: source/expectations/data_values/ut_compound_data_helper.pkb (hunk 4 above)
Comment (utplsql-travis) [inline source/expectations/data_values/ut_compound_data_helper.pkb]: Code Smell Code Smell: An illegal character with code point 10 was found in this literal. ([plsql:LiteralsNonPrintableCharactersCheck](https://sonarcloud.io/organizations/utplsql/rules#rule_key=plsql%3ALiteralsNonPrintableCharactersCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=utPLSQL&pullRequest=760&issues=AWY_N4w2nzfF1D4JsH2O&open=AWY_N4w2nzfF1D4JsH2O)

[ISSUE_KEY]: AWY_N4w2nzfF1D4JsH2O

**Comment by lwasylow** [comment]
That will be moved as part of include_matcher as some more thoughts has to be done into algorithm

**Pull request closed without merging**